syntax = "proto3";

package log_processor;

// 로그 프로세서 서비스
service LogProcessor {
  // 로그 파일을 파싱하고 Parquet으로 변환하여 업로드 (스트리밍 응답으로 진행상황 전달)
  rpc ProcessLogs(ProcessLogsRequest) returns (stream ProcessLogsProgress);
  
  // 작업 상태 조회
  rpc GetJobStatus(JobStatusRequest) returns (JobStatusResponse);
  
  // 버킷의 파일 목록 조회
  rpc ListFiles(ListFilesRequest) returns (ListFilesResponse);
}

// 로그 처리 요청
message ProcessLogsRequest {
  // 소스 버킷 이름
  string source_bucket = 1;
  
  // 소스 로그 파일 경로
  string source_path = 2;
  
  // 타겟 버킷 이름
  string target_bucket = 3;
  
  // Parquet이 저장될 경로
  string target_path = 4;
  
  // 로그 타입 (ufs, block, ufscustom)
  string log_type = 5;
  
  // 청크 크기 (옵션, 기본값: 100000)
  optional int32 chunk_size = 6;
}

// 로그 처리 진행 상황 (스트리밍)
message ProcessLogsProgress {
  // 작업 ID
  string job_id = 1;
  
  // 진행 단계
  ProgressStage stage = 2;
  
  // 진행 메시지
  string message = 3;
  
  // 진행률 (0-100)
  int32 progress_percent = 4;
  
  // 처리된 레코드 수
  int64 records_processed = 5;
  
  // 성공 여부 (완료시에만)
  optional bool success = 6;
  
  // 에러 메시지 (실패시에만)
  optional string error = 7;
  
  // 생성된 파일 경로들
  repeated string output_files = 8;
}

// 진행 단계
enum ProgressStage {
  STAGE_UNKNOWN = 0;
  STAGE_DOWNLOADING = 1;
  STAGE_PARSING = 2;
  STAGE_CONVERTING = 3;
  STAGE_UPLOADING = 4;
  STAGE_COMPLETED = 5;
  STAGE_FAILED = 6;
}

// 작업 상태 조회 요청
message JobStatusRequest {
  string job_id = 1;
}

// 작업 상태 응답
message JobStatusResponse {
  string job_id = 1;
  ProgressStage stage = 2;
  string message = 3;
  int32 progress_percent = 4;
  int64 records_processed = 5;
  bool is_completed = 6;
  optional bool success = 7;
  optional string error = 8;
}

// 파일 목록 조회 요청
message ListFilesRequest {
  string bucket = 1;
  string prefix = 2;
}

// 파일 목록 응답
message ListFilesResponse {
  repeated string files = 1;
}
